lib/AST/Type.cpp

```diff
@@ -1190,6 +1191,13 @@ CanType TypeBase::computeCanonicalType() {
     Result = BoundGenericType::get(BGT->getDecl(), parentTy, CanGenericArgs);
     break;
   }
+  case TypeKind::OpaqueTypeArchetype: {
+    // Canonicalize the substitution map.
+    OpaqueTypeArchetypeType *OAT = cast<OpaqueTypeArchetypeType>(this);
+    auto canonicalSubs = OAT->getSubstitutions().getCanonical();
+    Result = OpaqueTypeArchetypeType::get(OAT->getOpaqueDecl(), canonicalSubs);
```

Comment:
This is going to end up creating a new GSB instance and finalizing it and such. Can we canonicalize all of the individual pieces and call into the other, more direct constructor that won't do that redundant work?

Reply:
Hmm. There's another issue here. `ArchetypeType` is a bit special because they've always been considered canonical by construction. That's why the `ArchetypeType` constructor takes its `ASTContext` by reference rather than by pointer. Types that have non-canonical forms take `ASTContext` pointers where NULL indicates that the type is not already canonical. I think you'll need to change the `ArchetypeType` constructor to take the `ASTContext` as a pointer, and `OpaqueTypeArchetypeType::get` will determine whether it's forming the canonical type to decide whether to pass down the context or NULL.